istioctl/cmd/istioctl/meshdump.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"istio.io/istio/istioctl/pkg/kubernetes"
+	"istio.io/istio/pkg/log"
+
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/spf13/cobra"
+)
+
+var (
+	outMeshDumpFilename string
+
+	meshDumpCmd = &cobra.Command{
+		Use:   "mesh-dump",
+		Short: "Creates an archive of mesh information for debugging purposes [kube only]",
+		Long: `
+Creates an archive of mesh information required for debugging.
+`,
+		Example: `# Create archive of mesh information for debugging
+    istioctl mesh-dump
+`,
+		Aliases: []string{"md"},
+		RunE: func(c *cobra.Command, args []string) error {
+			kubeClient, err := newExecClient(kubeconfig, configContext)
+			if err != nil {
+				return err
+			}
+
+			filename := "istio-dump.tgz"
+			if outMeshDumpFilename != "" {
+				filename = outMeshDumpFilename
+			}
+
+			f, err := os.Create(filename)
+			if err != nil {
+				return err
+			}
+
+			zipper := gzip.NewWriter(f)
+			tarWriter := tar.NewWriter(zipper)
+			defer func() {
+				tarWriter.Close()
+				zipper.Close()
+				f.Close()
+			}()
+
+			var errs error
+			pilotADSz, err := kubeClient.AllPilotsDiscoveryDo(istioNamespace, "GET", "/debug/adsz", nil)
+			if err != nil {
+				log.Errorf("error retrieving adsz from Pilots: %v", err)
+			}
+			log.Debugf("Finished AllPilotsDiscoveryDo with %d pilots", len(pilotADSz))
+			for pilot, adsz := range pilotADSz {
```

Comment:
Is it worth parallelising this and executing each of the Envoy/Pilot calls concurrently in order to make them a little faster? Otherwise this command could take a while at scale.

If you do, make sure to use wait groups and not DoS the Kube api server!